google-cloud-storage/src/main/java/com/google/cloud/storage/StorageException.java

```diff
@@ -91,13 +95,30 @@ private static StorageException getStorageException(Throwable t) {
    * created with an unknown status code.
    */
   static BaseServiceException coalesce(Throwable t) {
-    // TODO: handle com.google.cloud.grpc.BaseGrpcServiceException et. al.
     if (t instanceof BaseServiceException) {
       return (BaseServiceException) t;
     }
     if (t.getCause() instanceof BaseServiceException) {
       return (BaseServiceException) t.getCause();
     }
+    if (t instanceof ApiException) {
+      ApiException apiEx = (ApiException) t;
+
+      // https://cloud.google.com/storage/docs/json_api/v1/status-codes
+      // https://cloud.google.com/apis/design/errors#http_mapping
+      // https://cloud.google.com/apis/design/errors#error_payloads
+      // TODO: flush this out more to wire through "error" and "details"
+      int httpStatusCode = 0;
+      StatusCode statusCode = apiEx.getStatusCode();
+      if (statusCode instanceof GrpcStatusCode) {
+        GrpcStatusCode gsc = (GrpcStatusCode) statusCode;
+        httpStatusCode =
+            BackwardCompatibilityUtils.grpcCodeToHttpStatusCode(gsc.getTransportCode());
+      }
+      String message = firstNonNull(() -> getCauseMessage(apiEx), apiEx::getMessage);
```

Comment:
one of these is guaranteed to be non null? if not we could risk throwing an illegal state exception instead of the StorageException we plan on constructing.

Reply:
You're correct, I've changed it so our base case is a generic message we construct.